File: baekjoon_all/10000+/boj_10464.java
// Solve 2024-02-20

// 백준에 제출할 때는 class 이름을 Main으로 설정해야 한다.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class boj_10464 {

    static int xorFromZero(int n) {
        int left = n % 2 == 0 ? n : 0;
        int right = (n % 4 == 1 || n % 4 == 2) ? 1 : 0;
        return left + right;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;
        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            st = new StringTokenizer(br.readLine(), " ");
            int start = Integer.parseInt(st.nextToken());
            int finish = Integer.parseInt(st.nextToken());
            sb.append(xorFromZero(start - 1) ^ xorFromZero(finish)).append("\n");
        }

        System.out.print(sb);
        br.close();
    }

}

File: baekjoon_all/10000+/boj_10464.cpp
// Solve 2024-02-20

#include <bits/stdc++.h>
using namespace std;

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL); // boj_15552.cpp
#define SETPRECISION(n) cout << fixed;cout.precision(n); // boj_1008.cpp
#define SIZE(v) (int)v.size()
#define ALL(v) v.begin(),v.end()
using ll = long long;

int xor_from_zero(int n) {
    int left = n % 2 == 0 ? n : 0;
    int right = (n % 4 == 1 || n % 4 == 2) ? 1 : 0;
    return left + right;
}

int main() {
    FASTIO;

    int t;
    cin >> t;

    while (t-- > 0) {
        int start, finish;
        cin >> start >> finish;
        cout << (xor_from_zero(start - 1) ^ xor_from_zero(finish)) << '\n';
    }

    return 0;
}
